src/main/java/org/myrobotlab/service/RasPi.java

```diff
@@ -376,6 +541,110 @@ public PinData publishPin(PinData pinData) {
     return pinData;
   }
 
+  public Map<Integer, Set<String>> publishScan() {
+    return validI2CAddresses;
+  }
+
+  public void read() {
+    log.debug("read task invoked");
+    List<PinData> pinArray = new ArrayList<>();
+    // load pin array
+    for (String pin : pinIndex.keySet()) {
+      PinDefinition pindef = pinIndex.get(pin);
+      if (pindef.isEnabled()) {
+        log.debug("pin {} enabled {}", pin, pindef.isEnabled());
+        int value = read(pin);
+        pindef.setValue(value);
+        PinData pd = new PinData(pin, value);
+        log.debug("pin data {}", pd);
+        pinArray.add(pd);
+      }
+    }
+
+    if (pinArray.size() > 0) {
+      PinData[] array = pinArray.toArray(new PinData[0]);
+      invoke("publishPinArray", new Object[] { array });
+    }
+  }
+
+  @Override
+  public int read(String pin) {
+
+    if (!pinIndex.containsKey(pin)) {
+      error("Pin %s not found", pin);
+      return -1;
+    }
+    PinDefinition pindef = pinIndex.get(pin);
+    GpioPinDigitalMultipurpose gpioPin = getGPIO(pin);
+    if (!gpioPin.isMode(PinMode.DIGITAL_INPUT)) {
+      pinMode(pin, "INPUT");
+    }
+    if (gpioPin.isLow()) {
+      pindef.setValue(0);
+      return 0;
+    } else {
+      pindef.setValue(1);
+      return 1;
+    }
+  }
+
+  @Override
+  public void reset() {
+    // TODO Auto-generated method stub
+    // reset pins/i2c devices/gpio pins
+  }
+
+  public void scan() {
+    scan(null);
+  }
+
+  public void scan(Integer busNumber) {
+
+    if (busNumber == null) {
+      busNumber = Integer.parseInt(bus);
+    }
+
+    try {
+
+      I2CBus bus = I2CFactory.getInstance(busNumber);
+
+      if (!validI2CAddresses.containsKey(busNumber)) {
+        validI2CAddresses.put(busNumber, new HashSet<>());
+      }
+
+      Set<String> addresses = validI2CAddresses.get(busNumber);
+
+      for (int i = 1; i < 128; i++) {
+        String hex = Integer.toHexString(i);
+        try {
+          I2CDevice device = bus.getDevice(i);
+          device.read();
+          if (!addresses.contains(hex)) {
+            addresses.add(hex);
+            info("found new i2c device %s", hex);
+          }
+        } catch (Exception ignore) {

```

Comment:
Can the set of caught exceptions be narrowed a bit? Right now it would end up ignoring critical exceptions like NPEs

Reply:
heh, totally fair 
... but I don't trust the i2c read to only throw an IOException.  If I separate the possibilities, I'd probably want the processing to be the same, and this would mean I have to copy/paste duplicate code.

I guess I'm just dealing with ptsd from seeing 
```java
} catch(X a) {
   ... do something ...
} catch(Y b) {
    ... do exactly the same something
} catch(Z c) {
    ... do exactly the same something

then someone has a fix for error handling behavior, but doesn't update all the other places....

Reply:
Java allows unions in catch() statements for exactly that reason:

```java
} catch (X | Y | Z e) {
   // Do thing for all exceptions in only one place
}
```